File: app/src/main/java/arun/com/chromer/activities/history/HistoryAdapter.java
/*
 * Chromer
 * Copyright (C) 2017 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.activities.history;

import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.util.AsyncListUtil;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.BitmapImageViewTarget;

import arun.com.chromer.R;
import arun.com.chromer.activities.BrowserInterceptActivity;
import arun.com.chromer.data.website.model.WebSite;
import arun.com.chromer.util.Utils;
import arun.com.chromer.views.PlaceholderLetterView;
import butterknife.BindView;
import butterknife.ButterKnife;
import rx.Observable;
import timber.log.Timber;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static arun.com.chromer.shared.Constants.EXTRA_KEY_FROM_OUR_APP;
import static arun.com.chromer.shared.Constants.EXTRA_KEY_SKIP_EXTRACTION;

/**
 * Created by Arunkumar on 06-03-2017.
 */
class HistoryAdapter extends RecyclerView.Adapter<HistoryAdapter.HistoryViewHolder> {
    private final AsyncListUtil<WebSite> asyncWebsiteList;
    private Cursor cursor = null;
    private final LinearLayoutManager linearLayoutManager;

    private final AsyncListUtil.DataCallback<WebSite> dataCallback = new AsyncListUtil.DataCallback<WebSite>() {
        @Override
        public int refreshData() {
            Timber.d("Refresh data");
            if (cursor == null || cursor.isClosed()) {
                return 0;
            }
            return cursor.getCount();
        }

        @Override
        public void fillData(WebSite[] data, int startPosition, int itemCount) {
            if (cursor != null) {
                for (int i = 0; i < itemCount; i++) {
                    cursor.moveToPosition(startPosition + i);
                    data[i] = WebSite.fromCursor(cursor);
                }
            }
        }
    };

    private final AsyncListUtil.ViewCallback viewCallback = new AsyncListUtil.ViewCallback() {
        @Override
        public void getItemRangeInto(int[] outRange) {
            outRange[0] = linearLayoutManager.findFirstVisibleItemPosition();
            outRange[1] = linearLayoutManager.findLastVisibleItemPosition();
        }

        @Override
        public void onDataRefresh() {
            Timber.d("onDataRefresh");
            notifyDataSetChanged();
        }

        @Override
        public void onItemLoaded(int position) {
            Timber.d("onItemLoaded, position %d", position);
            notifyItemChanged(position);
        }
    };

    HistoryAdapter(@NonNull LinearLayoutManager linearLayoutManager) {
        this.linearLayoutManager = linearLayoutManager;
        asyncWebsiteList = new AsyncListUtil<>(WebSite.class, 50, dataCallback, viewCallback);
    }

    @Override
    public HistoryViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new HistoryViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.activity_history_list_item_template, parent, false));
    }

    @Override
    public void onBindViewHolder(final HistoryViewHolder holder, int position) {
        final WebSite webSite = asyncWebsiteList.getItem(position);
        holder.bind(webSite);
    }

    @Override
    public void onViewRecycled(HistoryViewHolder holder) {
        super.onViewRecycled(holder);
        Glide.clear(holder.historyFavicon);
    }

    @Override
    public int getItemCount() {
        return asyncWebsiteList.getItemCount();
    }

    @Nullable
    WebSite getItemAt(final int position) {
        if (cursor != null) {
            cursor.moveToPosition(position);
            return WebSite.fromCursor(cursor);
        }
        return null;
    }

    @NonNull
    Observable<WebSite> getWebsiteAt(final int position) {
        return Observable.fromCallable(() -> getItemAt(position));
    }

    void setCursor(@Nullable Cursor cursor) {
        if (this.cursor != null) {
            try {
                this.cursor.close();
            } catch (Exception ignored) {
            }
        }
        this.cursor = cursor;
        refresh();
    }

    void onRangeChanged() {
        asyncWebsiteList.onRangeChanged();
    }

    void refresh() {
        asyncWebsiteList.refresh();
    }

    void cleanUp() {
        if (cursor != null) {
            cursor.close();
        }
    }

    class HistoryViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.history_title)
        TextView historyTitle;
        @BindView(R.id.history_favicon)
        ImageView historyFavicon;
        @BindView(R.id.history_subtitle)
        TextView historySubtitle;
        @BindView(R.id.history_amp)
        ImageView historyAmp;
        @BindView(R.id.history_placeholder)
        PlaceholderLetterView historyPlaceholder;


        HistoryViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            itemView.setOnClickListener(v -> {
                final int position = getAdapterPosition();
                final WebSite webSite = asyncWebsiteList.getItem(position);
                if (webSite != null && position != RecyclerView.NO_POSITION) {
                    final Intent intent = new Intent(itemView.getContext(), BrowserInterceptActivity.class);
                    intent.setData(Uri.parse(webSite.preferredUrl()));
                    intent.putExtra(EXTRA_KEY_FROM_OUR_APP, true);
                    itemView.getContext().startActivity(intent);
                }
            });

            historyAmp.setOnClickListener(v -> {
                final int position = getAdapterPosition();
                final WebSite webSite = asyncWebsiteList.getItem(position);
                if (webSite != null && position != RecyclerView.NO_POSITION) {
                    final Intent intent = new Intent(itemView.getContext(), BrowserInterceptActivity.class);
                    intent.setData(Uri.parse(webSite.ampUrl));
                    intent.putExtra(EXTRA_KEY_FROM_OUR_APP, true);
                    intent.putExtra(EXTRA_KEY_SKIP_EXTRACTION, true);
                    itemView.getContext().startActivity(intent);
                }
            });
        }

        public void bind(@Nullable WebSite webSite) {
            if (webSite == null) {
                historyTitle.setText(R.string.loading);
                historySubtitle.setText(R.string.loading);
                historyFavicon.setImageDrawable(null);
                historyAmp.setVisibility(GONE);
                Glide.clear(historyFavicon);
            } else {
                historyTitle.setText(webSite.safeLabel());
                historySubtitle.setText(webSite.preferredUrl());
                if (!TextUtils.isEmpty(webSite.faviconUrl)) {
                    Glide.with(itemView.getContext())
                            .load(webSite.faviconUrl)
                            .asBitmap()
                            .into(new BitmapImageViewTarget(historyFavicon) {
                                @Override
                                protected void setResource(Bitmap resource) {
                                    if (Utils.isValidFavicon(resource)) {
                                        showFavicon(resource);
                                    } else {
                                        showPlaceholder(webSite.safeLabel());
                                    }
                                }
                            });
                } else {
                    Glide.clear(historyFavicon);
                    showPlaceholder(webSite.safeLabel());
                }
                if (!TextUtils.isEmpty(webSite.ampUrl)) {
                    historyAmp.setVisibility(VISIBLE);
                } else {
                    historyAmp.setVisibility(GONE);
                }
            }
        }

        private void showPlaceholder(@NonNull String url) {
            historyFavicon.setImageDrawable(null);
            historyFavicon.setVisibility(GONE);
            historyPlaceholder.setVisibility(VISIBLE);
            historyPlaceholder.setPlaceHolder(url);
        }

        private void showFavicon(Bitmap resource) {
            historyPlaceholder.setVisibility(GONE);
            historyFavicon.setVisibility(VISIBLE);
            historyFavicon.setImageBitmap(resource);
        }
    }


}

File: app/src/main/java/arun/com/chromer/history/HistoryAdapter.kt
/*
 * Lynket
 *
 * Copyright (C) 2018 Arunkumar
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package arun.com.chromer.history

import android.app.Activity
import android.database.Cursor
import android.support.v7.util.AsyncListUtil
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import arun.com.chromer.Chromer
import arun.com.chromer.R
import arun.com.chromer.data.website.model.Website
import arun.com.chromer.extenstions.gone
import arun.com.chromer.extenstions.show
import arun.com.chromer.tabs.DefaultTabsManager
import arun.com.chromer.util.glide.GlideApp
import arun.com.chromer.util.glide.GlideRequests
import butterknife.BindView
import butterknife.ButterKnife
import timber.log.Timber

/**
 * Created by Arunkumar on 06-03-2017.
 */
internal class HistoryAdapter(
        activity: Activity,
        private val linearLayoutManager: LinearLayoutManager
) : RecyclerView.Adapter<HistoryAdapter.HistoryViewHolder>() {

    private val asyncWebsiteList: AsyncListUtil<Website>
    private var cursor: Cursor? = null
    private val glideRequests: GlideRequests
    private var tabsManager: DefaultTabsManager

    private val dataCallback = object : AsyncListUtil.DataCallback<Website>() {
        override fun refreshData(): Int {
            Timber.d("Refresh data")
            return if (cursor != null && !cursor!!.isClosed) {
                cursor!!.count
            } else 0
        }

        override fun fillData(data: Array<Website>, startPosition: Int, itemCount: Int) {
            if (cursor != null) {
                for (i in 0 until itemCount) {
                    cursor!!.moveToPosition(startPosition + i)
                    data[i] = Website.fromCursor(cursor!!)
                }
            }
        }
    }

    private val viewCallback = object : AsyncListUtil.ViewCallback() {
        override fun getItemRangeInto(outRange: IntArray) {
            outRange[0] = linearLayoutManager.findFirstVisibleItemPosition()
            outRange[1] = linearLayoutManager.findLastVisibleItemPosition()
        }

        override fun onDataRefresh() {
            Timber.d("onDataRefresh")
            notifyDataSetChanged()
        }

        override fun onItemLoaded(position: Int) {
            Timber.d("onItemLoaded, position %d", position)
            notifyItemChanged(position)
        }
    }

    init {
        asyncWebsiteList = AsyncListUtil(Website::class.java, 50, dataCallback, viewCallback)
        glideRequests = GlideApp.with(activity)
        tabsManager = (activity.application as Chromer).appComponent.defaultTabsManager()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): HistoryViewHolder {
        return HistoryViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.activity_history_list_item_template, parent, false))
    }

    override fun onBindViewHolder(holder: HistoryViewHolder, position: Int) {
        val website = asyncWebsiteList.getItem(position)
        holder.bind(website)
    }

    override fun onViewRecycled(holder: HistoryViewHolder?) {
        super.onViewRecycled(holder)
        glideRequests.clear(holder?.historyFavicon)
    }

    override fun getItemCount(): Int {
        return asyncWebsiteList.itemCount
    }

    fun getItemAt(position: Int): Website? {
        return asyncWebsiteList.getItem(position)
    }

    fun setCursor(cursor: Cursor?) {
        if (this.cursor != null && this.cursor != cursor) {
            try {
                this.cursor?.close()
            } catch (e: Exception) {
                Timber.e(e)
            }
        }
        this.cursor = cursor
        refresh()
    }

    fun onRangeChanged() {
        asyncWebsiteList.onRangeChanged()
    }

    private fun refresh() {
        asyncWebsiteList.refresh()
    }

    fun cleanUp() {
        cursor?.close()
    }

    internal inner class HistoryViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        @BindView(R.id.history_title)
        @JvmField
        var historyTitle: TextView? = null
        @BindView(R.id.history_favicon)
        @JvmField
        var historyFavicon: ImageView? = null
        @BindView(R.id.history_subtitle)
        @JvmField
        var historySubtitle: TextView? = null
        @BindView(R.id.history_amp)
        @JvmField
        var historyAmp: ImageView? = null

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener {
                val position = adapterPosition
                val website = asyncWebsiteList.getItem(position)
                if (website != null && position != RecyclerView.NO_POSITION) {
                    tabsManager.openUrl(itemView.context, website)
                }
            }

            historyAmp?.setOnClickListener {
                val position = adapterPosition
                val website = asyncWebsiteList.getItem(position)
                if (website != null && position != RecyclerView.NO_POSITION) {
                    tabsManager.openUrl(itemView.context, Website.Ampify(website))
                }
            }
        }

        fun bind(website: Website?) {
            if (website == null) {
                historyTitle?.setText(R.string.loading)
                historySubtitle?.setText(R.string.loading)
                historyFavicon?.setImageDrawable(null)
                historyAmp?.visibility = GONE
                glideRequests.clear(historyFavicon!!)
            } else {
                historyTitle?.text = website.safeLabel()
                historySubtitle?.text = website.preferredUrl()
                GlideApp.with(itemView.context)
                        .load(website)
                        .into(historyFavicon!!)

                if (website.hasAmp()) {
                    historyAmp?.show()
                } else {
                    historyAmp?.gone()
                }
            }
        }
    }
}